ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/WsAgentWebSocketConnectionTracker.java

```diff
@@ -0,0 +1,160 @@
+ * WsAgentServerStoppedEvent} when the connection is lost. Prompts to restart a workspace when
+ * connection is lost, but workspace is running.
+ *
+ * @author Roman Nikitenko
+ */
+@Singleton
+public class WsAgentWebSocketConnectionTracker {
+  private static final int HEART_BEAT_PERIOD =
+      WS_AGENT_TRACK_CONNECTION_PERIOD + WS_AGENT_TRACK_CONNECTION_PERIOD / 2;
+
+  private EventBus eventBus;
+  private AppContext appContext;
+  private DialogFactory dialogFactory;
+  private ActionManager actionManager;
+  private RequestTransmitter requestTransmitter;
+  private CoreLocalizationConstant localizationConstant;
+
+  private String devMachineName;
+
+  private final Timer heartBeatTimer =
+      new Timer() {
+        @Override
+        public void run() {
+          onWsAgentConnectionIsLost();
+        }
+      };
+
+  @Inject
+  public WsAgentWebSocketConnectionTracker(
+      EventBus eventBus,
+      AppContext appContext,
+      DialogFactory dialogFactory,
+      ActionManager actionManager,
+      RequestTransmitter transmitter,
+      RequestHandlerConfigurator requestHandler,
+      CoreLocalizationConstant localizationConstant) {
+    this.appContext = appContext;
+    this.eventBus = eventBus;
+    this.actionManager = actionManager;
+    this.requestTransmitter = transmitter;
+    this.dialogFactory = dialogFactory;
+    this.localizationConstant = localizationConstant;
+
+    requestHandler
+        .newConfiguration()
+        .methodName(WS_AGENT_TRACK_CONNECTION_HEARTBEAT)
+        .noParams()
+        .noResult()
+        .withConsumer(s -> restartTimer());
+
+    eventBus.addHandler(
+        BasicIDEInitializedEvent.TYPE,
+        event ->
+            appContext
+                .getWorkspace()
+                .getDevMachine()
+                .flatMap(machine -> machine.getServerByName(SERVER_WS_AGENT_HTTP_REFERENCE))
+                .map(ServerImpl::getStatus)
+                .filter(RUNNING::equals)
+                .ifPresent(it -> initialize()));
+
+    eventBus.addHandler(WsAgentServerRunningEvent.TYPE, e -> initialize());
+    eventBus.addHandler(WsAgentServerStoppedEvent.TYPE, e -> unsubscribe());
+  }
+
+  private void initialize() {
+    Optional<MachineImpl> devMachine = appContext.getWorkspace().getDevMachine();
+    devMachineName = devMachine.isPresent() ? devMachine.get().getName() : "";
+
+    requestTransmitter
+        .newRequest()
+        .endpointId(WS_AGENT_JSON_RPC_ENDPOINT_ID)
+        .methodName(WS_AGENT_TRACK_CONNECTION_SUBSCRIBE)
+        .noParams()
+        .sendAndSkipResult();
+
+    restartTimer();
+  }
+
+  private void restartTimer() {
+    heartBeatTimer.cancel();
+    heartBeatTimer.schedule(HEART_BEAT_PERIOD);
+  }
+
+  private void onWsAgentConnectionIsLost() {
+    eventBus.fireEvent(new WsAgentServerStoppedEvent(devMachineName));
+
+    Workspace workspace = appContext.getWorkspace();
+    if (workspace == null || workspace.getStatus() != WorkspaceStatus.RUNNING) {
```

Comment:
IDE initialization has been improved a lot in Che 6.
Current workspace info set in `AppContext` during IDE initialization. And if IDE can't reach a workspace on loading - IDE initialization fails.
So, in Che 6, checking `workspace == null` is redundant since `appContext.getWorkspace()` can't return `null`.
If it isn't so, please ping me.